ddl/index.go

```diff
@@ -325,30 +325,54 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
 	return errors.Trace(err)
 }
 
-func fetchRowColVals(txn kv.Transaction, t table.Table, handle int64, indexInfo *model.IndexInfo) (
-	kv.Key, []types.Datum, error) {
-	// fetch datas
-	cols := t.Cols()
-	colMap := make(map[int64]*types.FieldType)
-	for _, v := range indexInfo.Columns {
-		col := cols[v.Offset]
-		colMap[col.ID] = &col.FieldType
+func (d *ddl) fetchRowColVals(txn kv.Transaction, t table.Table, handles []int64, indexInfo *model.IndexInfo) (
+	[]*indexRecord, error) {
+	// Through handles access to get all row keys.
+	handlesLen := len(handles)
+	rowKeys := make([]kv.Key, 0, handlesLen)
+	idxRecords := make([]*indexRecord, handlesLen)
+	for i, h := range handles {
+		rowKey := tablecodec.EncodeRecordKey(t.RecordPrefix(), h)
+		rowKeys = append(rowKeys, rowKey)
+		idxRecords[i] = &indexRecord{handle: h, key: rowKey}
```

Comment:
Line 336 only appends an object already exist, line 337 needs to append a new object.